### PR TITLE
qa/tasks: assert on pg status with a timeout

### PR DESCRIPTION
**Pull request by tchaikov**
Fixes: http://tracker.ceph.com/issues/19594
Signed-off-by: Kefu Chai <kchai@redhat.com>

### CODE REVIEW

**Comment by tchaikov** [comment]
being tested at http://pulpito.ceph.com/kchai-2017-04-18_13:04:04-rados-wip-kefu-testing---basic-mira/

**Comment by jcsp** [comment]
Hmm, why did this work when the command was going to the mon, but need a sleep now that it's in mgr?  The mon had the same stats period didn't it?

**Comment by tchaikov** [comment]
good question. the osd_mon_report_interval_min is also 5. i have no idea why we don't have this problem before.

**Comment by liewegas** [comment]
I think this is a band-aid.  The problem is the stats delay (mon stats also lag a bit too).  The problem is that there are weird implicit timing assumptoins in the tests that once one condition is true (e.g., wait_for_healthy) then the pg stats will reflect that.  It used to be true because both conditions were coming out of the mon, but now the mon and mgr info is disconnected, and we get occasionaly failures due to that.

We *also* have the problem that an ill-timed mgr failure-over may kick the pg stats back in time a bit too.

I'm worried that we need to sort these cases out individually by fixing the tests.  I'm also concerned that this function is called from lots of placing and making it sleep 10s every time is going to be a problem...

**Comment by tchaikov** [comment]
so shall we decrease the interval for qa tests?

**Comment by tchaikov** [comment]
we need a "pg flush" command to force all OSDs to send their pg stats to  mgr right away.

**Comment by liewegas** [comment]
Two ideas:

1. do something like the osd tell flush_pg_stats thing so that we block until the stats go to the mgr
2. accept that pg stats are "eventually consistent" and wrap every test/assertion about them in a loop with a timeout.

**Comment by tchaikov** [comment]
i vote for option 2 =)

**Comment by tchaikov** [comment]
being tested at http://pulpito.ceph.com/kchai-2017-04-19_09:02:59-rados-master---basic-mira/

**Comment by tchaikov** [comment]
@jcsp i think we failed to schedule a scrub because the "last_scrub_stamp" returned by the first and the second "pg dump" calls in `CephManager.do_pg_scrub()` are different, because the different ctimes of pg-create messages before luminous and after luminous.  i included c52d035 in this PR to address this.

File: qa/tasks/ceph_manager.py [hunk 1]
```diff
@@ -1326,10 +1326,10 @@ def get_pg_primary(self, pool, pgnum):
...
         get primary for pool, pgnum (e.g. (data, 0)->0
         """
-        pgstr = self.get_pgid(pool, pgnum)
-        stats = self.get_single_pg_stats(pgstr)
-        assert stats is not None
-        return int(stats['acting'][0])
+        def primary(stats):
+            assert stats is not None
+            return int(stats['acting'][0])
+        return self.with_pg(pool, pgnum, primary)
```
Comment (tchaikov) [inline qa/tasks/ceph_manager.py]: this change will conflict with #14559

**Comment by tchaikov** [comment]
@liewegas fixed and  repushed.

**Comment by tchaikov** [comment]
changelog

- dropped the change of `get_pg_primary()` which would otherwise conflict with #14559
- add more sleep values in the `wait_for_pg_stats()` decorator

**Comment by tchaikov** [comment]
tested at http://pulpito.ceph.com/kchai-2017-04-20_01:40:17-rados-wip-19594-kefu---basic-mira/

**Pull request merged**